Added parse tree equivalence understandings

mr-martian opened the pull request:
I have added a simple parse tree modifier that makes
```3 * 4``` match ```4 * 3```
and
```i = i + 3``` match ```i += 3``` and vice versa.
I have done my best to test it, though I have not checked anything much more complicated than the above.

pamelafox commented:
Neat! I am worried about changes to structuredjs however, as I'm not sure how it'd affect the challenges. We have regression tests for each challenge that we currently run manually in the browser when we're adjusting the structured tests for a challenge. I think we'd ideally want to automatically re-run all those tests and make sure they still produced the same rule. It's possible those could all be run outside our browser framework, and I could provide them to you. They do have some live-editor dependency as well.
Would you be interested to try to get those running?

mr-martian commented:
Sure, I'd just need some instructions for how to run them and I could probably get it done in the next few days.

mr-martian commented:
I ran ```structuredjs/browser-test/structuredjs_test.html```, here's a screenshot:
![screenshot from 2015-03-12 12 55 32](https://cloud.githubusercontent.com/assets/5350777/6624529/dc9d298c-c8b7-11e4-8971-5982ee8e34a7.png)
Are there any other tests you were reffering to that you would like me to do?

pamelafox commented:
Sorry, what I mean is that every JS challenge on Khan Academy uses StructuredJS for grading, and each one has its own is-this-correct-tests and regression tests.
I put an example here:
https://gist.github.com/pamelafox/9631d54cb0da304b3b61
That's for:
https://www.khanacademy.org/computing/computer-science/algorithms/binary-search/p/challenge-binary-search

It's technically out of the scope of structuredjs, but ideally, I don't want to push large changes to structuredjs and update live-editor with the new structuredjs unless I'm very confident that it doesn't affect any of our challenge graders negatively.
So you could come up with a way to automatically run those in live-editor and confirm the regression tests still work for all the challenges.
Or you could persuade me that this is a backwards-compatible change, which I'll admit, it does seem to be (and it seems like a pretty great one).

mr-martian commented:
If you give me some idea how, I am willing to run the tests.
However, since you listed convincing you as an option:
I would say that it is backwards compatible because it only runs ```restructureTree()``` when ```checkMatchTree()``` has tried everything else and would otherwise return false. Thus it only applies my change when it otherwise would have failed, so everything that would have passed before should still pass now.

pamelafox commented:
Okay, that does seem reasonable. Could you add a test in structuredjs that verifies that 2*3 would match 3*2 and i = i + 3 match i += 3 ? That way, if we come to rely on that behavior in our challenge graders, somebody doesn't accidentally remove that.

mr-martian commented:
I just discovered that the test on lines 729-736 contains a match of ```tree += 30 + 50 + 70;``` to ```_ += $a + $b;``` and requires that ```$a > $b``` (intentionally fails to match). Should I add an option for not re-ordering variables or comment out the test? (since my addition will inevitably cause this test to fail)

pamelafox commented:
Ah, yes, I don't think we would have that test in a grader... although I suppose we might if we were checking string concatenation order. Is your equivalence only for Numbers or also on strings?

mr-martian commented:
I forgot that + also applies to strings and didn't check for type.
restructureTree() now checks for options.orderMatters before applying the commutative property.
I modified the test to use this option (https://github.com/mr-martian/structuredjs/blob/master/tests.js#L733)

Anyway, I added 4 tests. They and all the old ones are working properly.

kevinb7 commented:
This is really awesome!

pamelafox commented on structured.js:
```diff
@@ -505,12 +506,53 @@
             // Recursively check for matches
             if ((_.isArray(currTree[key]) &&
-                    checkNodeArray(currTree[key], toFind, peersToFind, wVars, matchResults, options)) ||
+                    checkNodeArray(currTree[key], toFind, peersToFind, wVars, matchResults, options, false)) ||
```
Hm, does this one need false passed to it? It doesn't look like you modified checkNodeArray.

pamelafox commented on structured.js:
```diff
@@ -478,7 +478,8 @@
      *     toFind (and on the same level as toFind).
      */
-    function checkMatchTree(currTree, toFind, peersToFind, wVars, matchResults, options) {
+    function checkMatchTree(currTree, toFind, peersToFind, wVars, matchResults, options, recursing) {
```
It looks like the jsdoc has gotten outdated for this function, can you update it?

pamelafox commented on structured.js:
```diff
@@ -505,12 +506,53 @@
                 return matchResults;
             }
         }
+        if (!recursing) {
```
Hm. It seems strange to me that this function seems to act recursively when recursing is false. Can you add comments that further explain how this works?

pamelafox commented on structured.js:
```diff
@@ -505,12 +506,53 @@
+    }
+    
+    /*
+        Applies simple transformations to the parse tree to reattempt matching
```
Can you copy the * style of the above comment? (*s on each line of the comment)

pamelafox commented on structured.js:
```diff
@@ -505,12 +506,53 @@
+    
+    /*
+        Applies simple transformations to the parse tree to reattempt matching
+        Takes an argument list identical to checkMatchTree() above
```
Can you add a simple example to this comment as well, like you had in the PR, like "this expression is turned into this expression"?

pamelafox commented on structured.js:
```diff
@@ -505,12 +506,53 @@
+                                operator: currTree.operator.slice(0,-1),
+                                left: currTree.left,
+                                right: currTree.right}};
+            }
```
Move else if to this line

pamelafox commented on tests.js:
```diff
@@ -1839,6 +1840,36 @@ var altVarCallbacks = function(){
     });
 };
 
+var commutativity = function(){
```
Thanks for the tests!

mr-martian commented:
I believe I have dealt with all but 2 of your line notes.
Concerning the first and third:
I passed false in because I dislike using default variables that I have not explicitly labeled as optional.
As for ```!recursing```, I went back and forth on whether it meant "we are recursing" or "continue recursing". It is probably the former at this point simply because that's what it was when I got it working.

mr-martian commented:
I have replaced ```!recursing``` with ```modify```, which doesn't seem to have equally comprehensible, contradictory meanings in that way.

pamelafox commented:
Thanks for the changes, I'm going to merge this!
